### ISSUE TITLE
[16.0] BUG: onchange in many2one field

### ISSUE THREAD

**Issue by igoroliveira-ax4b** [odoo/odoo#172858]
### Odoo Version

- [ ] 15.0
- [X] 16.0
- [ ] 17.0
- [ ] Other (specify)

### Steps to Reproduce

- Create any field many2one
- create a function with a decorator @api.onchange('many2one_field')
- On browser, select any value at many2one field and save
- Try to change the value

When I change the value in the many2one field, the value remains at the value before the change
This problem occurs in any module.

**Comment by intero-chz** [comment]
Could you add your python code and xml? I doubt such a dramatic mistake, because many developers would justifiably cry or freak out in the issues long ago.

**Comment by igoroliveira-ax4b** [comment]
@intero-chz 

It's not a complex code


from odoo import models, fields, api

```
class Requirements(models.Model):
    _name = 'geff.requirements'
    _description = 'Requirements'

    # other fields...
    requirement_type_id = fields.Many2one(
        comodel_name='geff.requirement_type',
        string='Requirement Type')

    @api.onchange('requirement_type_id')
    def _onchange_requirement_type_id(self):
        for rec in self:
            remove = []
            lines = []
            for line in rec.document_line_ids:
                remove.append([2,line.id])            
            for item in rec.requirement_type_id.document_type:
                lines.append([0,0,{'requirement_doc_type_id': item.id}])
            rec.update({'document_line_ids': remove})
            rec.update({'document_line_ids': lines})
```




in xml file is not complex
just a form with a line

```
...
<field name="requirement_type_id" />
...
```

**Comment by intero-chz** [comment]
That looks like a change on a one2many field. IIRC the framework is limited in this respect.

**Comment by igoroliveira-ax4b** [comment]
no, this code try to change an many2one field in a common form

**Comment by intero-chz** [comment]
Yes you change a many2one in the form, but your onchange function is changing a one2many field (`document_line_ids`), That's what i meant.

**Comment by igoroliveira-ax4b** [comment]
I have another example, with the same problem:

```
    @api.onchange("asset_profile_id")
    def _onchange_asset_profile_id(self):
        if self.asset_profile_id.account_asset_id:
            self.account_id = self.asset_profile_id.account_asset_id
```

This code check if the field "account_asset_id" is filled in asset_profile_id record then write it on account_id field in current form, without involve O2M fields

**Comment by intero-chz** [comment]
Then there must be another error somewhere. A third-party module may be causing this error.

If you create and save a new product variant with the name “Test” in the current Runbot for Odoo 16, you can then change the UoM to “Dozens”. Odoo automatically changes the PO UoM (via an onchange).
